README.md
# pretty_cron

Provides a human-readable format for cron expressions.

Actual docs probably coming in the future.


HISTORY.rst
.. :changelog:

History
-------

1.2.0 (2018-08-22)
------------------

* **Retired** in favour of cron_descriptor_. This project is no longer maintained.

.. _cron_descriptor: https://pypi.org/project/cron_descriptor/

1.1.0 (2018-08-19)
------------------

* Now always returns ``unicode`` on Python 2 for consistency.

1.0.2 (2016-05-03)
------------------

* Now supports expressions with multiple weekdays and ordinal days - thanks
  @jbondia.

1.0.1 (2016-02-09)
------------------

* Now interprets day 7 as Sunday as well as 0, like Linux crontab parsers -
  thanks @vetyy.
* Now supports expressions with multiple months separated by commas - thanks
  @MerreM.

1.0.0 (2015-07-28)
------------------

* First release on PyPI, featuring ``prettify_cron`` function.


tests.py
import unittest

from pretty_cron import prettify


class PrettyCronTest(unittest.TestCase):
    def test_yearly(self):
        self.assertEqual(prettify("0 0 1 1 *"),
                         "At 00:00 on the 1st of January")

    def test_one_day_in_month(self):
        self.assertEqual(prettify("0 0 1 * *"),
                         "At 00:00 on the 1st of every month")

    def test_every_day_in_month(self):
        self.assertEqual(prettify("12 15 * 1 *"),
                         "At 15:12 every day in January")

    def test_every_specific_day_in_month(self):
        self.assertEqual(prettify("0 0 * 1 1"),
                         "At 00:00 on every Monday in January")

    def test_weekly(self):
        self.assertEqual(prettify("0 0 * * 0"),
                         "At 00:00 every Sunday")

    def test_monthly_and_weekly(self):
        self.assertEqual(prettify("0 0 1 * 1"),
                         "At 00:00 on the 1st of every month and every Monday")

    def test_every_specific_day_in_month_and_weekly(self):
        self.assertEqual(
            prettify("0 0 1 1 1"),
            "At 00:00 on the 1st of January and on every Monday in January"
        )

    def test_daily(self):
        self.assertEqual(prettify("0 0 * * *"), "At 00:00 every day")

    def test_hourly(self):
        self.assertEqual(prettify("0 * * * *"),
                         "At 0 minutes past every hour of every day")

    def test_minutely(self):
        self.assertEqual(prettify("* 5 * * *"),
                         "Every minute between 05:00 and 05:59 every day")

    def test_continuous(self):
        self.assertEqual(prettify("* * * * *"), "Every minute of every day")

    def test_unsupported(self):
        self.assertEqual(prettify("* */6 * * *"), "* */6 * * *")

if __name__ == '__main__':
    unittest.main()


test_prettify.py
import unittest

from pretty_cron import prettify


class PrettyCronTest(unittest.TestCase):
    def test_yearly(self):
        self.assertEqual(prettify("0 0 1 1 *"),
                         "At 00:00 on the 1st of January")

    def test_one_day_in_month(self):
        self.assertEqual(prettify("0 0 1 * *"),
                         "At 00:00 on the 1st of every month")

    def test_every_day_in_month(self):
        self.assertEqual(prettify("12 15 * 1 *"),
                         "At 15:12 every day in January")

    def test_every_specific_day_in_month(self):
        self.assertEqual(prettify("0 0 * 1 1"),
                         "At 00:00 on every Monday in January")

    def test_weekly(self):
        self.assertEqual(prettify("0 0 * * 0"),
                         "At 00:00 every Sunday")

    def test_monthly_and_weekly(self):
        self.assertEqual(prettify("0 0 1 * 1"),
                         "At 00:00 on the 1st of every month and every Monday")

    def test_every_specific_day_in_month_and_weekly(self):
        self.assertEqual(
            prettify("0 0 1 1 1"),
            "At 00:00 on the 1st of January and on every Monday in January"
        )

    def test_daily(self):
        self.assertEqual(prettify("0 0 * * *"), "At 00:00 every day")

    def test_hourly(self):
        self.assertEqual(prettify("0 * * * *"),
                         "At 0 minutes past every hour of every day")

    def test_minutely(self):
        self.assertEqual(prettify("* 5 * * *"),
                         "Every minute between 05:00 and 05:59 every day")

    def test_continuous(self):
        self.assertEqual(prettify("* * * * *"), "Every minute of every day")

    def test_unsupported(self):
        self.assertEqual(prettify("* */6 * * *"), "* */6 * * *")


tests/test_prettify.py
# -*- encoding:utf-8 -*-
from __future__ import absolute_import, division, print_function, unicode_literals

import pytest

from pretty_cron import prettify_cron as pc


def test_yearly():
    assert pc("0 0 1 1 *") == "At 00:00 on the 1st of January"


def test_one_day_in_month():
    assert pc("0 0 1 * *") == "At 00:00 on the 1st of every month"


def test_one_day_in_month_2nd():
    assert pc("0 0 2 * *") == "At 00:00 on the 2nd of every month"


def test_one_day_in_month_11th():
    assert pc("0 0 11 * *") == "At 00:00 on the 11th of every month"


def test_one_day_in_month_21st():
    assert pc("0 0 21 * *") == "At 00:00 on the 21st of every month"


def test_every_day_in_month():
    assert pc("12 15 * 1 *") == "At 15:12 every day in January"


def test_every_day_in_months():
    assert (
        pc("12 15 * 1,12 *") ==
        "At 15:12 every day in January and December"
    )


def test_every_specific_day_in_month():
    assert pc("0 0 * 1 1") == "At 00:00 on every Monday in January"


def test_every_pair_of_days_in_month():
    assert (
        pc("0 0 * 1 1,2") ==
        "At 00:00 on every Monday and Tuesday in January"
    )


def test_every_three_days_in_month():
    assert (
        pc("0 0 * 1 1,2,3") ==
        "At 00:00 on every Monday, Tuesday and Wednesday in January"
    )


def test_every_specific_day_in_months():
    assert (
        pc("0 0 * 1,2 1") ==
        "At 00:00 on every Monday in January and February"
    )


def test_every_specific_day_in_months_and_then_more():
    assert (
        pc("0 0 * 1,2,4,5 1") ==
        "At 00:00 on every Monday in January, February, April and May"
    )


def test_every_multiple_days_in_multiple_months():
    assert (
        pc("0 0 * 1,2,4,5 1,2,3") ==
        "At 00:00 on every Monday, Tuesday and Wednesday "
        "in January, February, April and May"
    )


def test_weekly():
    assert pc("0 0 * * 0") == "At 00:00 every Sunday"


def test_day_7_is_sunday():
    assert pc("0 0 * * 7") == "At 00:00 every Sunday"


def test_monthly_and_weekly():
    assert (
        pc("0 0 1 * 1") ==
        "At 00:00 on the 1st of every month and every Monday"
    )


def test_every_specific_day_in_month_and_weekly():
    assert (
        pc("0 0 1 1 1") ==
        "At 00:00 on the 1st of January and on every Monday in January"
    )


def test_every_specific_day_in_months_and_weekly():
    assert (
        pc("0 0 1 1,2 1") ==
        "At 00:00 on the 1st of January and February and on every Monday "
        "in January and February"
    )


def test_every_pair_of_days_in_multiple_months_and_weekly():
    assert (
        pc("0 0 1,23 1,2 1") ==
        "At 00:00 on the 1st and 23rd of January and February and on "
        "every Monday in January and February"
    )


def test_every_pair_of_days_in_multiple_months_and_twice_weekly():
    assert (
        pc("0 0 1,23 1,2 1,4") ==
        "At 00:00 on the 1st and 23rd of January and February and on "
        "every Monday and Thursday in January and February"
    )


def test_every_pair_of_days_in_multiple_months_and_three_days_weekly():
    assert (
        pc("0 0 1,23 1,2 1,4,6") ==
        "At 00:00 on the 1st and 23rd of January and February and on "
        "every Monday, Thursday and Saturday in January and February"
    )


def test_every_multiple_days_in_multiple_months_and_weekly():
    assert (
        pc("0 0 1,2,3,4 1,2 1") ==
        "At 00:00 on the 1st, 2nd, 3rd and 4th of January and February "
        "and on every Monday in January and February"
    )


def test_every_specific_day_in_months_and_more_and_weekly():
    assert (
        pc("0 0 1 1,2,3 1") ==
        "At 00:00 on the 1st of January, February and March and on every "
        "Monday in January, February and March"
    )


def test_daily():
    assert pc("0 0 * * *") == "At 00:00 every day"


def test_hourly():
    assert pc("0 * * * *") == "At 0 minutes past every hour of every day"


def test_minutely():
    assert (
        pc("* 5 * * *") == "Every minute between 05:00 and 05:59 every day"
    )


def test_continuous():
    assert pc("* * * * *") == "Every minute of every day"


def test_unsupported():
    assert pc("* */6 * * *") == "* */6 * * *"


def test_invalid_unchanged():
    assert pc('* * * * * *') == '* * * * * *'


def test_nonsense_unchanged():
    assert pc('Lalalala') == 'Lalalala'


# Tests for "step values" in crontabs...
@pytest.mark.parametrize('crontab,result', [
    ('4/21 * * * *', 'Every 21st minute from 4 through 59 every day'),
    ('0/30 3 * * *', 'Every 30th minute from 0 through 59 past hour 3 every day'),
    ('10-50/15 * * * 7', 'Every 15th minute from 10 through 50 every Sunday'),
    ('2-46/11 * * * *', 'Every 11th minute from 2 through 46 every day'),
    ('* 2-6/2 * * *', 'Every minute of every 2nd hour from 2 through 6 every day'),
    ('0 2-6/2 * * *', 'At the start of every 2nd hour from 2 through 6 every day'),
    ('17 2-6/2 * * *', 'Every 17th minute of every 2nd hour from 2 through 6 every day'),
])
def test_step_values_mins_hours(crontab, result):
    assert pc(crontab) == result


def test_step_values_for_mins_and_hours_raises():
    with pytest.raises(ValueError):
        pc('0/30 4/10 * * *')


runtests.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import absolute_import, division, print_function, unicode_literals

import subprocess
import sys

import pytest
import six
from flake8.main import main as flake8_main
from libmodernize.main import main as libmodernize_main


CODE_PATHS = [
    'setup.py',
    'runtests.py',
    'pretty_cron',
    'tests',
]


def main():
    try:
        sys.argv.remove('--nolint')
    except ValueError:
        run_lint = True
    else:
        run_lint = False

    try:
        sys.argv.remove('--lintonly')
    except ValueError:
        run_tests = True
    else:
        run_tests = False

    if run_tests:
        exit_on_failure(tests_main())

    if run_lint:
        exit_on_failure(run_flake8())
        exit_on_failure(run_modernize())
        exit_on_failure(run_isort())

        # Broken on 2.7.9 due to http://bugs.python.org/issue23063
        if sys.version_info[:3] != (2, 7, 9):
            exit_on_failure(run_setup_py_check())


def tests_main():
    return pytest.main()


def run_flake8():
    print('Running flake8 code linting')
    try:
        original_argv = sys.argv
        sys.argv = ['flake8'] + CODE_PATHS
        did_fail = False
        flake8_main()
    except SystemExit:
        did_fail = True
    finally:
        sys.argv = original_argv

    print('flake8 failed' if did_fail else 'flake8 passed')
    return did_fail


def run_modernize():
    print('Running modernize checks')
    try:
        orig_stdout = getattr(sys, 'stdout')
        out = six.StringIO()
        setattr(sys, 'stdout', out)
        libmodernize_main([
            '-x', 'lib2to3.fixes.fix_has_key'
        ] + CODE_PATHS)
    finally:
        sys.stdout = orig_stdout
    output = out.getvalue()
    print(output)
    ret = len(output)
    print('modernize failed' if ret else 'modernize passed')
    return ret


def run_isort():
    print('Running isort check')
    return subprocess.call([
        'isort', '--recursive', '--check-only', '--diff',
        '-a', 'from __future__ import absolute_import, division',
        '-a', 'from __future__ import print_function, unicode_literals',
    ] + CODE_PATHS)


def run_setup_py_check():
    print('Running setup.py check')
    return subprocess.call([
        'python', 'setup.py', 'check',
        '-s', '--restructuredtext', '--metadata'
    ])


def exit_on_failure(ret, message=None):
    if ret:
        sys.exit(ret)


if __name__ == '__main__':
    main()


pretty_cron/api.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import, division, print_function, unicode_literals

import datetime
import re

# regex to detect cron step expressions, such as 5-30/10 = "every 10 from 5 through 30"
step_pattern = re.compile(r'^(?P<start>\d{1,2})(?:-(?P<end>\d{1,2}))?/(?P<step>\d{1,2})$')

# regex to capture various crontab tokens
cron_pattern = re.compile(r'^(?P<fullexpr>(?P<allexpr>\*)'
                          r'|(?P<valexpr>\d{1,2})'
                          r'|(?P<rangeexpr>(?P<from>\d{1,2})-(?P<to>\d{1,2}))'
                          r'|(?P<stepexpr>(?P<start>\d{1,2})(?:-(?P<end>\d{1,2}))?/(?P<step>\d{1,2}))'
                          r'|(?P<listexpr>\d{1,2}(?:,\d{1,2})+))$')


def prettify_cron(expression):
    """
    Converts the given string cron expression into a pretty, human-readable,
    English description of what it means. If the string is not a valid cron
    expression, or it includes features not currently supported, it is returned
    as-is.
    """
    pieces = []
    for i, piece in enumerate(expression.split(" ")):
        # support comma-separated values for ordinal days, weekdays and months
        if i in (2, 3, 4) and ',' in piece:
            try:
                piece = tuple(int(p) for p in piece.split(','))
            except ValueError:
                # non-integers in comma-separated list aren't supported yet -
                # return as-is
                return expression
        elif piece != '*' and not (step_pattern.match(piece) and i in (0, 1)):
            try:
                piece = int(piece)
            except ValueError:
                # */2 and other cron expressions aren't supported yet - return
                # as-is
                return expression
        pieces.append(piece)

    try:
        minute, hour, month_day, month, week_day = pieces
    except ValueError:
        # More or fewer pieces than expected - return as-is
        return expression

    date = _pretty_date(month_day, month, week_day)
    time = _pretty_time(minute, hour)

    return " ".join(
        x for x in (time, date) if x
    )


def _pretty_date(month_day, month, week_day):
    if month_day == "*" and week_day == "*":
        pretty_date = "every day"

        if month != '*':
            pretty_date += " in {0}".format(_human_month(month))
    else:
        if month_day != "*":
            month_day_date = "on the {0}".format(_ordinal(month_day))
        else:
            month_day_date = ""

        if week_day != "*":
            week_day_date = "every {0}".format(_human_week_day(week_day))
        else:
            week_day_date = ""

        if month_day_date:
            if month != '*':
                month_day_date += " of {0}".format(_human_month(month))
            else:
                month_day_date += " of every month"

        if week_day_date and month != "*":
            week_day_date = "on {0} in {1}".format(
                week_day_date,
                _human_month(month)
            )

        pretty_date = " and ".join(
            x for x in (month_day_date, week_day_date) if x
        )

    return pretty_date


def _human_month(month):
    if isinstance(month, tuple):
        months = month
    else:
        months = (month,)

    return _human_list([
        datetime.date(2014, m, 1).strftime('%B')
        for m in months
    ])


def _human_list(listy):
    """
    Combines a list of strings into a string representing the list in plain
    English, i.e.
    >>> _human_list(["a"])
    "a"
    >>> _human_list(["a", "b"])
    "a and b"
    >>> _human_list(["a", "b", "c"])
    "a, b and c"
    """
    if len(listy) == 1:
        return listy[0]

    rest, penultimate, ultimate = listy[:-2], listy[-2], listy[-1]
    if rest:
        return ", ".join(rest) + ", {} and {}".format(penultimate, ultimate)
    return "{} and {}".format(penultimate, ultimate)


_WEEKDAYS = {
    0: "Sunday",
    1: "Monday",
    2: "Tuesday",
    3: "Wednesday",
    4: "Thursday",
    5: "Friday",
    6: "Saturday",
    7: "Sunday",
}


def _human_week_day(day):
    if isinstance(day, tuple):
        days = day
    else:
        days = (day,)

    return _human_list([_WEEKDAYS[d] for d in days])


_ORDINAL_SUFFIXES = {
    1: 'st',
    2: 'nd',
    3: 'rd'
}


def _ordinal(num):
    if isinstance(num, tuple):
        nums = num
    else:
        nums = (num,)

    ordinal_days = []
    for d in nums:
        if 10 <= (d % 100) < 20:
            suffix = 'th'
        else:
            suffix = _ORDINAL_SUFFIXES.get(d % 10, 'th')
        ordinal_days.append(str(d) + suffix)

    return _human_list(ordinal_days)


def _pretty_time_with_steps(minute, hour):
    if minute['stepexpr']:
        start = int(minute['start'])
        end = int(minute['end'] or 59)
        step = _ordinal(int(minute['step']))

        hour_expr = ""
        if hour['allexpr'] != '*':
            hour_expr = " past hour {0}".format(hour['valexpr'])
        return "Every {step} minute from {m_start} through {m_end}{hour_expr}".format(
            step=step,
            m_start=start,
            m_end=end,
            hour_expr=hour_expr)

    start = int(hour['start'])
    end = int(hour['end'] or 23)
    step = _ordinal(int(hour['step']))

    if minute['allexpr']:
        min_expr = "Every minute"
    else:
        minval = int(minute['valexpr'])
        min_expr = "Every {0} minute".format(_ordinal(minval)) if minval != 0 else "At the start"
    return "{min_expr} of every {step} hour from {m_start} through {m_end}".format(
        step=step,
        m_start=start,
        m_end=end,
        min_expr=min_expr)


def _pretty_time(minute, hour):
    min_groups, hour_groups = (m and m.groupdict() for m in (cron_pattern.match(str(p)) for p in (minute, hour)))
    # see if either the min or hour are a step expression
    step_count = sum(bool(g['stepexpr']) for g in (min_groups, hour_groups))
    if step_count > 0:
        if step_count == 2:
            raise ValueError('Step expressions only work for either minutes or hours, but not both at this time.')
        return _pretty_time_with_steps(min_groups, hour_groups)

    if minute != "*" and hour != "*":
        the_time = datetime.time(hour=hour, minute=minute)
        pretty_time = "At {0}".format(the_time.strftime("%H:%M"))

    elif minute != "*" and hour == '*':
        pretty_time = "At {0} minutes past every hour of".format(minute)

    elif minute == "*" and hour != '*':
        start_time = datetime.time(hour=hour)
        end_time = datetime.time(hour=hour, minute=59)

        pretty_time = "Every minute between {0} and {1}".format(
            start_time.strftime("%H:%M"),
            end_time.strftime("%H:%M"),
        )
    else:
        pretty_time = "Every minute of"

    return pretty_time
